subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/ImplementationSnapshot.java

```diff
@@ -26,10 +28,39 @@
  * the classloader hierarchy by its hash code.
  */
 public class ImplementationSnapshot implements ValueSnapshot {
+    private static final String GENERATED_LAMBDA_CLASS_SUFFIX = "$$Lambda$";
+
     private final String typeName;
     private final HashCode classLoaderHash;
 
-    public ImplementationSnapshot(String typeName, @Nullable HashCode classLoaderHash) {
+    public static ImplementationSnapshot of(Class<?> type, ClassLoaderHierarchyHasher classLoaderHasher) {
+        return of(determineImplementationName(type.getName(), type.isSynthetic()), classLoaderHasher.getClassLoaderHash(type.getClassLoader()));
+    }
+
+    public static ImplementationSnapshot of(String typeName, @Nullable HashCode classLoaderHash) {
+        return new ImplementationSnapshot(typeName, classLoaderHash);
+    }
+
+    public static ImplementationSnapshot withDeterministicClassName(String className, @Nullable HashCode classLoaderHash) {
+        return of(determineImplementationName(className, true), classLoaderHash);
+    }
+
+    private static String determineImplementationName(String className, boolean mayBeLambda) {
+        if (mayBeLambda && className.contains(GENERATED_LAMBDA_CLASS_SUFFIX)) {
+            SingleMessageLogger.nagUserWith(
+                "Java lambda is used as an input.",
+                "Gradle can only track the lambda used in your code with some loss of precision that may lead to some changes going undetected.",
```

Comment:
Why do we reduce the precision if we are deprecating this anyway? Why not just deprecate lambdas as actions and keep the old behavior of rerunning more often than necessary?

Reply:
And I'm saying "deprecate", because logging a message like this is essentially equivalent to forbidding it for anyone who cares about their build warnings.

Reply:
The goal is to motivate users to avoid using lambdas in `doLast()` calls.

We chose to reduce precision and keep the cache working because that will cover 99.9% of use cases as intended. A false positive cache hit requires a rather special kind of setup where a warning could help the user figure out what went wrong.

I would personally prefer a simple `Logger.warn()`-type of warning (at least until we have a better API for non-deprecation warnings) to make it "lighter." In either case avoiding the warning should be simple: just change the lambda into an anonymous inner class.

Another option would be to explicitly disable caching if we detected a lambda input, however, in that case up-to-date checks would also produce incorrect results (again, with pretty low probability, but not absurdly low).

Leaving the behavior as is now is not an option, because currently a cacheable task with a lambda input will produce a cache miss on every build, resulting in a store of its outputs that is plain overhead.

Reply:
I don't see how any of those behaviors are better than the other. I think we should just deprecate lambdas as actions if we can't deal with them correctly.

Reply:
@oehme We are now disabling caching (build cache & up-to-date checks) when a lambda is used as an input. So a task which uses a lambda as an input will never be up-to-date as opposed to being sometimes up-to-date.

Reply:
It was not about the nagging: we shouldn't allow things to be stored in cache that we know will never produce a cache hit, thus only wasting the time.